lock/src/test/java/org/infinispan/lock/ClusteredLockSplitBrainTest.java

```diff
@@ -59,118 +79,128 @@ public void testLockUseAfterPartitionWithoutMajority() throws Throwable {
+      ClusteredLock lock5 = clm5.get(LOCK_NAME);
+
+      splitCluster(new int[]{0, 1, 2, 3}, new int[]{4, 5});
+
+      asList(lock0, lock1, lock2, lock3).forEach(lock -> {
+         assertTrue("Lock acquisition should be true " + lock, await(lock.tryLock(29, TimeUnit.SECONDS)
+               .thenApply(tryLockRequest -> {
+                  if (tryLockRequest) {
+                     await(lock.unlock());
+                  }
+                  return tryLockRequest;
+               }).exceptionally(ex -> {
+                  fail("Should not be failing from majority partition " + lock + " " + ex.getMessage());
+                  return null;
+               })));
+
+      });
+
+      assertFailureFromMinorityPartition(lock4);
+      assertFailureFromMinorityPartition(lock5);
+   }
 
-      splitCluster(new int[]{0, 1, 2}, new int[]{3});
+   @Test
+   public void testAutoReleaseIfLockIsAcquiredFromAMinorityPartition() throws Throwable {
 
-      await(lock0.tryLock(5, TimeUnit.SECONDS)
-            .thenRun(() -> {
-               lock0.unlock();
-            }).exceptionally(ex -> {
-               fail("Should not be failing from majority partition");
-               return null;
-            }));
+      ClusteredLockManager clm0 = EmbeddedClusteredLockManagerFactory.from(getCacheManagers().get(0));
+      ClusteredLockManager clm1 = EmbeddedClusteredLockManagerFactory.from(getCacheManagers().get(1));
+      ClusteredLockManager clm2 = EmbeddedClusteredLockManagerFactory.from(getCacheManagers().get(2));
 
-      await(lock1.tryLock(5, TimeUnit.SECONDS)
-            .thenRun(() -> {
-               lock1.unlock();
-            }).exceptionally(ex -> {
-               fail("Should not be failing from majority partition");
-               return null;
-            }));
+      assertTrue(clm0.defineLock(LOCK_NAME));
 
-      await(lock2.tryLock(5, TimeUnit.SECONDS)
-            .thenRun(() -> {
-               lock2.unlock();
-            }).exceptionally(ex -> {
-               fail("Should not be failing from majority partition");
-               return null;
-            }));
+      ClusteredLock lock0 = clm0.get(LOCK_NAME);
+      ClusteredLock lock1 = clm1.get(LOCK_NAME);
+      ClusteredLock lock2 = clm2.get(LOCK_NAME);
 
-      assertFailureFromMinorityPartition(lock3);
+      await(lock0.tryLock());
+      assertTrue(await(lock0.isLockedByMe()));
+
+      splitCluster(new int[]{0}, new int[]{1, 2, 3, 4, 5});
+
+      CompletableFuture<Boolean> tryLock1Result = lock1.tryLock(1, TimeUnit.SECONDS);
+      CompletableFuture<Boolean> tryLock2Result = lock2.tryLock(1, TimeUnit.SECONDS);
+
+      assertTrue("Just one of the locks has to work", await(tryLock1Result) ^ await(tryLock2Result));
+
+      assertFailureFromMinorityPartition(lock0);
    }
 
    @Test
-   public void testAutoReleaseIfLockIsAcquiredFromAMinorityPartition() throws Throwable {
+   public void testTryLocksBeforeSplitBrain() throws Throwable {
```

Comment:
So does this tests always pass?

Reply:
It fails a lot on the topology change issue. I want to work on that with @danberindei